homeassistant/components/zwave_js/light.py

```diff
@@ -321,6 +324,9 @@ async def _async_set_brightness(
         if brightness is None:
             # Level 255 means to set it to previous value.
```

Comment:
Does this actually set it to the previous value? I'm asking, because it'd be nice that if the last time HA knew it was on, that it assumed the brightness would be that value the next time it was turned on by HA. Right now, it seems like this will have no brightness value (but show that it's on) until it actually gets a report from the zwave device.

Reply:
The last brightness level is stored in the device. This is simply the value used to instruct the device to turn on to the last level.

Reply:
That is to say, my specific issue is that I don't get _any_ reports from my device for ~4 seconds, in which it reports its final brightness. I didn't 100% follow the other connect thread talking about it.

Reply:
@kpine got it. What do you think about updating the UI to optimistically be the last known brightness in this scenario?

Reply:
I don't have much of an opinion, but it doesn't sound unreasonable to me. Storing the last brightness in the entity data is already done by some integrations, so it's not a novel idea.

Reply:
It'd be nice. It should be correct the vast majority of the time, and it will self-correct after a bit, anyway. Compare that to now: when it's _always_ wrong by saying the brightness is 0.

This is the experience you get when you get a SmartThings/Hubitat hub and connect via the community's mqtt bridge. It reports it right away and is almost always right. Makes everything feel a lot more responsive.

Reply:
And yeah, like you said, the other option is to send last known brightness instead of `255`, cuz then we'd _always_ be right.

Reply:
I think that's a bad idea, it won't always be right. The device knows what the last level was.

Reply:
If the device knows what the last value was, why does it way 3 seconds to tell anybody about it? If you send a brightness of 87, the device _immediately_ responds with "my brightness is 87," even before the light is actually at 87. But if you send that 255 thing, it seems like it's waiting a while cuz it's not sure... or something.

What about that other solution where we always send the last known value?

Reply:
> If the device knows what the last value was, why does it way 3 seconds to tell anybody about it?

The problem here is usually these affected devices never report a new value when issued via a Z-Wave command. So ask your device manufacturer why. The value update you're seeing is probably a refresh from the driver (which would be 5 seconds). Some devices report changes unsolicited, others don't. The PR here is addressing the latter. It also addresses user transitions > 2 seconds, which have the same behavior.

> If you send a brightness of 87, the device immediately responds with "my brightness is 87," even before the light is actually at 87. But if you send that 255 thing, it seems like it's waiting a while cuz it's not sure... or something.

The device is no responding, the driver is sending an optimistic value, it's an intentional design to fix this same issue but for known level values. If want to know more about this topic you can read https://github.com/zwave-js/node-zwave-js/issues/452. I don't think we need to re-hash it in this PR.

I think it's debatable whether you choose to optimistically report on but no brightness (more accurate because you're unsure) or the last brightness (chance it's wrong). I personally don't have an opinion. The former is simpler to implement.

> What about that other solution where we always send the last known value?

I don't agree with this. Let the device handle this behavior that it's designed for. Since you've already made a comparison to how SmartThings works, I'll do the same 😉. SmartThings does not set the level to a specific "last known" value when you turn on the light, it [sets the 255 value](https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/cbfccd9058762ab9765c01216976d730bf267352/devicetypes/smartthings/dimmer-switch.src/dimmer-switch.groovy#L208) to let the device handle it, meaning it's the same as HA. This is if you're using the standard "Dimmer Switch" device handler.

Just my opinions of course. 😉

Reply:
I looked over what ST + MQTT Bridge does, and it looks like it just handles state vs level, similar to what's being proposed in the PR here. I thought they sent the brightness when you turn it on, but it doesn't at all. I was wrong there.

But that means something in HA is just setting it to its last known level in lieu of a brightness level. Perhaps something specific to the MQTT Light code.

Reply:
Also, I wanna say thank you for taking the time to discuss this with me. This issue has been around for years, and I know lots has been said. I'm not trying to re-hash on beat-to-death topics. I am excited to see some work to help for a better experience for those of us in this situation. 

I've been using ST + MQTT Bridge (and later Hubitat + MQTT Bridge) and have had a great experience there, but am hoping to put everything in the HA fold for convenience/performance, etc, and want to be able to have the same user experience, so I'm just trying to help, but realize I'm probably mostly wasting your time and energy... so I appreciate you humoring me. :) And especially the efforts of the firstof9 and all of you here.